tap_pipedrive/tap.py

```diff
@@ -244,17 +244,43 @@ def execute_stream_request(self, stream):
         return self.execute_request(stream.endpoint, params=params)
 
     def execute_request(self, endpoint, params=None):
-        headers = {"User-Agent": self.config["user-agent"]}
-        _params = {
-            "api_token": self.config["api_token"],
-        }
-        if params:
-            _params.update(params)
-
         url = "{}/{}".format(BASE_URL, endpoint)
-        logger.debug("Firing request at {} with params: {}".format(url, _params))
+        logger.debug("Firing request at {} with params: {}".format(url, params))
 
-        return requests.get(url, headers=headers, params=_params)
+        # remove this case when all customers have authed with oauth
+        if "api_token" in self.config:
+            params = {"api_token": self.config["api_token"], **params}
+            headers = {"User-Agent": self.config["user-agent"]}
+            response = requests.get(url, headers=headers, params=params)
+        else:
+            headers = {
+                "User-Agent": self.config["user-agent"],
+                "Authorization": f"Bearer {self.access_token}",
+            }
+            response = requests.get(url, headers=headers, params=params)
+
+            if response.status_code == 401:
+                # attempt to refresh access token
+                self.refresh_access_token()
+                headers["Authorization"] = f"Bearer {self.access_token}"
+                response = requests.get(url, headers=headers, params=params)
+
+        return response
+
+    def refresh_access_token(self):
+        # remove this check when all customers have authed with oauth
+        if "api_token" not in self.config:
+            payload = {
+                "grant_type": "refresh_token",
+                "refresh_token": self.config["refresh_token"],
+                "client_id": self.config["client_id"],
+                "client_secret": self.config["client_secret"],
+            }
+            resp = requests.post(OAUTH_URL + "/token", data=payload)
+            resp.raise_for_status()
+            if not resp:
```

Comment:
is the `if not resp` not redundant with the `raise_for_status()` just above it? I believe it tests the same things, but I am not sure

Reply:
👍 its true, no need for this